bee-common/bee-packable/src/packable/prefix.rs

```diff
@@ -104,102 +73,80 @@ impl<T, B: Bounded> Default for VecPrefix<T, B> {
-                self.inner
-            }
-        }
+    fn try_from(vec: Vec<T>) -> Result<Self, Self::Error> {
+        B::try_from(vec.len())?;
 
-        impl<T, $($generics)*> core::ops::Deref for VecPrefix<T, $bounded> {
-            type Target = Vec<T>;
+        Ok(Self {
+            inner: vec,
+            bounded: PhantomData,
+        })
+    }
+}
 
-            fn deref(&self) -> &Self::Target {
-                &self.inner
-            }
+impl<T, B> Packable for VecPrefix<T, B>
+where
+    T: Packable,
+    B: Bounded + Packable,
+    <B::Bounds as TryInto<B>>::Error: Debug,
+    <B as TryFrom<usize>>::Error: Debug,
+{
+    type UnpackError = UnpackPrefixError<T::UnpackError, B::UnpackError>;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        // The length of any dynamically-sized sequence must be prefixed.
+        // This unwrap is fine, since we have already validated the length in `try_from`.
+        B::try_from(self.len()).unwrap().pack(packer)?;
+
+        for item in self.iter() {
+            item.pack(packer)?;
         }
 
-        impl<T: Packable, $($generics)*> Packable for VecPrefix<T, $bounded> {
-            type UnpackError = $unpack_err;
-
-            fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
-                // The length of any dynamically-sized sequence must be prefixed.
-                // This unwrap is fine, since we have already validated the length in `try_from`.
-                <$ty>::try_from(self.len()).unwrap().pack(packer)?;
-
-                for item in self.iter() {
-                    item.pack(packer)?;
-                }
-
-                Ok(())
-            }
-
-            fn unpack<U: Unpacker, const VERIFY: bool>(
-                unpacker: &mut U,
-            ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
-                // The length of any dynamically-sized sequence must be prefixed.
-                let len: $ty = <$bounded>::unpack::<_, VERIFY>(unpacker)
-                    .map_packable_err($map_err)?
-                    .into();
-
-                let mut inner = Vec::with_capacity(len as usize);
-
-                for _ in 0..len {
-                    let item = T::unpack::<_, VERIFY>(unpacker).coerce()?;
-                    inner.push(item);
-                }
-
-                Ok(VecPrefix {
-                    inner,
-                    bounded: PhantomData,
-                })
-            }
-        }
-    };
-}
+        Ok(())
+    }
+
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        // The length of any dynamically-sized sequence must be prefixed.
+        let len = B::unpack::<_, VERIFY>(unpacker)
+            .map_packable_err(UnpackPrefixError::Prefix)?
+            .into();
+
+        let mut inner = Vec::with_capacity(len.try_into().unwrap_or(0));
```

Comment:
Maybe i'm just tired but what exactly would it mean if that was failing ?

Reply:
It's a bit of a corner case that ends up failing almost all the time anyway:  If `B = u64` and we were in a 32-bit platform then it might for any  `len > u32::MAX`. If that happens you create a `0` capacity vector and then you start pushing elements of type `T` to it until you run out of memory.

The only way OOM doesn't happen is if `T` has size-zero. So, in principle you can decode `len` values of type `T` without reaching memory exhaustion.